core/src/main/java/fj/data/Stream.java

```diff
@@ -386,13 +386,7 @@ public final void foreachDoEffect(final Effect1<A> f) {
    * @return A new stream after performing the map, then final join.
    */
   public final <B> Stream<B> bind(final F<A, Stream<B>> f) {
-    return map(f).foldLeft((accumulator, element) -> {
-        Stream<B> result = accumulator;
-        for (B single : element) {
-            result = result.cons(single);
-        }
-        return result;
-    }, Stream.<B>nil()).reverse();
+    return foldRight(h -> (t -> f.f(h).append(t)), nil());
   }
```

Comment:
I am concerned about using O(n) stack space here using foldRight given it is a common operation.